File: src/Domain/Tournaments/Grain.fs
namespace Orleans.Tournament.Domain.Tournaments
open Microsoft.Extensions.Logging
open Orleans.Tournament.Domain.Abstractions
open Orleans.Tournament.Domain.Abstractions.Grains
open Orleans.Tournament.Domain.Helpers
open Orleans.Tournament.Domain.Teams
open System.Threading.Tasks
open Orleans

type ITournamentGrain =
    inherit IGrainWithGuidKey
    // Commands
    abstract CreateAsync: CreateTournament -> Task
    abstract AddTeamAsync: AddTeam -> Task
    abstract StartAsync: StartTournament -> Task
    abstract SetMatchResultAsync: SetMatchResult -> Task

type TournamentGrain(logger : ILogger<TournamentGrain>) =
    inherit EventSourcedGrain<TournamentState>(
        new StreamOptions(MemoryProvider, TournamentNamespace),
        logger)

    interface ITournamentGrain with
        member x.CreateAsync cmd =
            match Rules.TournamentDoesNotExists x.State with
            | Ok _ -> x.PersistPublishAsync (TournamentCreated.From cmd)
            | Error error -> x.EmitErrorAsync error cmd

        member x.AddTeamAsync cmd =
            // There is no problem on validating on the controller using the
            // read side tables, but this validation should be done on the domain
            // as this is the entry point of truth
            let teamGrain = x.GrainFactory.GetGrain<ITeamGrain>(cmd.TeamId)
            // TODO: Any other form of handling this, in an Fsharp way, deadlocks
            let task = Task.Run(fun _ -> teamGrain.TeamExistAsync().Result)

            match
                Rules.TournamentExists x.State <&>
                Rules.TeamExistsForward task.Result <&>
                Rules.TeamIsNotAdded x.State cmd.TeamId <&>
                Rules.LessThanEightTeams x.State with
              | Ok _ -> x.PersistPublishAsync (TeamAdded.From cmd)
              | Error error -> x.EmitErrorAsync error cmd

        member x.StartAsync cmd =
            match
                Rules.TournamentExists x.State <&>
                Rules.TournamentIsNotStarted x.State <&>
                Rules.EightTeamsToStartTournament x.State with
              | Ok _ -> x.PersistPublishAsync (TournamentStarted.From cmd x.State.Teams )
              | Error error -> x.EmitErrorAsync error cmd

        member x.SetMatchResultAsync cmd =
            match
                Rules.TournamentExists x.State <&>
                Rules.TournamentIsStarted x.State <&>
                Rules.MatchIsNotDraw cmd.MatchInfo.MatchSummary <&>
                Rules.MatchExistsAndIsNotPlayed x.State cmd.MatchInfo with
              | Ok _ -> x.PersistPublishAsync (MatchResultSet.From cmd)
              | Error error -> x.EmitErrorAsync error cmd

    member x.EmitErrorAsync (error : BusinessErrors) (cmd : ITraceable) =
        base.PublishErrorAsync((int error), error.ToString(), cmd.TraceId, cmd.InvokerUserId)

File: src/DomainC/Tournaments/Grain.cs
using Orleans;
using Tournament.Domain.Abstractions;
using Tournament.Domain.Abstractions.Events;
using Tournament.Domain.Abstractions.Grains;
using Tournament.Domain.Teams;

namespace Tournament.Domain.Tournaments;

public interface ITournamentGrain : IGrainWithGuidKey
{
    // Commands
    Task CreateAsync(CreateTournament cmd);

    Task AddTeamAsync(AddTeam cmd);

    Task StartAsync(StartTournament cmd);

    Task SetMatchResultAsync(SetMatchResult cmd);
}

public class TournamentGrain : EventSourcedGrain<TournamentState>, ITournamentGrain
{
    public TournamentGrain()
        : base(new StreamConfig(Constants.InMemoryStream, Constants.TournamentNamespace))
    { }

    public async Task CreateAsync(CreateTournament cmd)
    {
        var task = State.TournamentDoesNotExists() switch
        {
            Results.Unit => PersistPublishAsync(new TournamentCreated(cmd.Name, cmd.TournamentId, cmd.TraceId, cmd.InvokerUserId)),
            Results x => PublishErrorAsync(new ErrorHasOccurred((int)x, nameof(CreateTournament), cmd.TraceId, cmd.InvokerUserId))
        };

        await task;
    }

    public async Task AddTeamAsync(AddTeam cmd)
    {
        // Check if the team already exists
        var teamGrain = GrainFactory.GetGrain<ITeamGrain>(cmd.TeamId);
        var exists = await teamGrain.TeamExistAsync();
        if (!exists)
            await PublishErrorAsync(new ErrorHasOccurred((int)Results.TeamDoesNotExist, nameof(AddTeam), cmd.TraceId, cmd.InvokerUserId));

        // Check other preconditions
        var result = ResultsUtil.Eval(
            State.TournamentExists(),
            State.TeamIsNotAdded(cmd.TeamId),
            State.LessThanEightTeams());

        var task = result switch
        {
            Results.Unit => PersistPublishAsync(new TeamAdded(cmd.TeamId, cmd.TournamentId, cmd.TraceId, cmd.InvokerUserId)),
            Results x => PublishErrorAsync(new ErrorHasOccurred((int)x, nameof(AddTeam), cmd.TraceId, cmd.InvokerUserId))
        };

        await task;
    }

    public async Task StartAsync(StartTournament cmd)
    {
        var result = ResultsUtil.Eval(
            State.TournamentExists(),
            State.TournamentDidNotStart(),
            State.EightTeamsToStartTournament());

        var task = result switch
        {
            Results.Unit => PersistPublishAsync(new TournamentStarted(cmd.TournamentId, State.Teams, cmd.TraceId, cmd.InvokerUserId)),
            Results x => PublishErrorAsync(new ErrorHasOccurred((int)x, nameof(TournamentStarted), cmd.TraceId, cmd.InvokerUserId))
        };

        await task;
    }

    public async Task SetMatchResultAsync(SetMatchResult cmd)
    {
        var result = ResultsUtil.Eval(
            State.TournamentExists(),
            State.TournamentStarted(),
            State.MatchIsNotDraw(cmd.Match.MatchResult!),
            State.MatchExistsAndIsNotPlayed(cmd.Match));

        var task = result switch
        {
            Results.Unit => PersistPublishAsync(new MatchResultSet(cmd.TournamentId, cmd.Match, cmd.TraceId, cmd.InvokerUserId)),
            Results x => PublishErrorAsync(new ErrorHasOccurred((int)x, nameof(SetMatchResult), cmd.TraceId, cmd.InvokerUserId))
        };

        await task;
    }
}